Add `hasParameter` and `hasNoParameter` assertion methods for URIs and URLs

benblank opened the pull request:
This allows assertions in the style of:
 
    assertThat(new URI("http://foo.com/?bar=baz")).hasParameter("bar");

Additionally, support is available for a few edge cases I've seen in the wild; specifically, parameters with no value (e.g. `?foo&bar`) as well as multiple parameters with the same name.  Therefore, it can be meaningful to write an assertion like:

    assertThat(new URI("http://foo.com/?bar&bar=baz")).hasParameter("bar", null).hasParameter("bar", "baz");

#### Check List:
* Fixes : NA
* Unit tests : YES
* Javadoc with a code example (API only) : YES

joel-costigliola commented:
Good idea, thanks for the PR. WIll review it soon

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractUrlAssert.java:
```diff
@@ -291,4 +291,104 @@ public S hasNoUserInfo() {
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // This assertion succeeds:
+   * assertThat(new URL("http://www.helloworld.org/index.html")).hasParameter("happy");
```
this is a valid a code example for `hasNoParameter` except that it shows `hasParameter` :p

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractUrlAssert.java:
```diff
@@ -291,4 +291,104 @@ public S hasNoUserInfo() {
+   * assertThat(new URL("http://www.helloworld.org/index.html")).hasParameter("happy", "very");
+   * assertThat(new URL("http://www.helloworld.org/index.html?happy")).hasParameter("happy", "very");
+   * assertThat(new URL("http://www.helloworld.org/index.html?happy=very")).hasParameter("happy", null);
+   * 
```
same comment no `hasNoParameter` used here

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractUriAssert.java:
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
+   * Use {@code null} to indicate an absent value (e.g. {@code foo&bar}) as opposed to an empty value (e.g.
+   * {@code foo=&bar=}).
+   * <p>
+   * Examples:
```
use `hasNoParameter` in the examples

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractUriAssert.java:
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
+   * The value of the parameter is not checked.
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // This assertion succeeds:
```
use `hasNoParameter` in the examples

joel-costigliola commented on src/main/java/org/assertj/core/error/uri/ShouldHaveParameter.java:
```diff
@@ -0,0 +1,141 @@
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_MISSING = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut was absent";
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_NO_VALUE = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut had no value";
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_WRONG = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut had value:%n  <%s>";
+  private static final String SHOULD_HAVE_NO_PARAMETER_BUT_NO_VALUE = "%nExpecting:%n  <%s>%nto not have parameter:%n  <%s>%nbut was present with no value";
```
I feel that `not to have a parameter` is better than `to not have parameter`

joel-costigliola commented on src/main/java/org/assertj/core/internal/Uris.java:
```diff
@@ -81,4 +89,91 @@ public void assertHasUserInfo(AssertionInfo info, URI actual, String expected) {
+    List<String> values = parameters.get(name);
+
+    if (!values.contains(value)) {
+      throw failures.failure(info, shouldHaveParameter(actual, name, value, values.get(0)));
```
are we sure values can't be empty ?
If we display parameter's values, we should display them all between `[]`, if there is only one, stick with the actual message.

joel-costigliola commented on src/main/java/org/assertj/core/internal/Uris.java:
```diff
@@ -81,4 +89,91 @@ public void assertHasUserInfo(AssertionInfo info, URI actual, String expected) {
+    if (parameters.containsKey(name)) {
+      List<String> values = parameters.get(name);
+
+      throw failures.failure(info, shouldHaveNoParameter(actual, name, values.get(0)));
```
same comment here on displaying multiple values.

benblank commented:
* **Bad examples** — Well, that's embarrassing!  Fixed.
* **Message phrasing** — No objection.  Changed.
* **Multiple values** — Good idea.  For single values, the messages now show it verbatim, but multiple values are shown "array style".  I've added a few tests to cover these cases, as well.

PascalSchumacher commented on src/main/java/org/assertj/core/api/AbstractUriAssert.java:
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // This assertion succeeds:
+   * assertThat(new URI("http://www.helloworld.org/index.html")).hasParameter("happy");
```
@benblank I guess you missed this line when fixing the javadoc examples?

joel-costigliola commented:
Thanks a lot @benblank, I have integrated your work with (I believe) few improvements and an additional assertion : `hasNoParameters()`.

I'm closing this PR but feel free to comment my changes if you feel they can be improved.
